treeherder/push_health/similar_jobs.py

```diff
@@ -0,0 +1,63 @@
+from itertools import groupby
+
+from django.db.models import Q
+from django.forms.models import model_to_dict
+
+from treeherder.model.models import Job
+
+job_fields = [
+    'id',
+    'machine_platform',
+    'option_collection_hash',
+    'job_type',
+    'result',
+    'failure_classification'
+]
+
+
+def set_matching_passed_jobs(failures, push):
+
+    failed_jobs = {}
+    for failure in failures:
+        jobs = {job['id']: job for job in failure['failJobs']}
+        failed_jobs.update(jobs)
+
+    # Need to OR for each type of job that failed
+    query_conditions = []
+    for job in failed_jobs.values():
+        query_conditions.append(Q(
+            option_collection_hash=job['option_collection_hash'],
+            machine_platform=job['machine_platform'],
+            job_type=job['job_type'],
+        ))
+    query = query_conditions.pop()
+    for condition in query_conditions:
+        query |= condition
+
+    passing_jobs = Job.objects.filter(
+        push=push,
+        result='success'
+    ).filter(query).select_related('job_type')
+
+    #
+    # Group the passing jobs into groups based on their platform, option and job_type
+    #
+    # convert from ORM objects to dicts for when we return this object
+    passing_job_dicts = [model_to_dict(job, fields=job_fields) for job in passing_jobs]
+    sorted_passing_jobs = sorted(passing_job_dicts, key=get_job_key)
+    passing_job_map = {key: list(group) for key, group in groupby(sorted_passing_jobs, get_job_key)}
+
+    #
+    # Assign matching passing jobs to the test failures
+    #
+    for failure in failures:
+        # If the passing jobs has a list matching this failure's jobs, then add them in
+        if len(failure['failJobs']):
+            # A failure will have the same job_key for all jobs in this push, so use the first one
+            job_key = get_job_key(failure['failJobs'][0])
+            if job_key in passing_job_map:
+                failure['passJobs'] = passing_job_map[job_key]
```

Comment:
does this get returned anywhere or do we update the passed in failures data structure?

Reply:
This sets the ``passJobs`` key in the ``failures`` object that was passed in, 
which is then returned from the API.